File: Task1B/tshirt.cpp
#include "tshirt.h"

// Default constructor
TShirt::TShirt() {
    this->collar = "";
    this->graphics = false;
}

// Parameterised constructor
TShirt::TShirt(string name, int price, Supplier* supplier, int volume, int productID, string size, string brand, string material, string collar, bool graphics) : Clothes(name, price, supplier, volume, productID, size, brand, material) {
    this->collar = collar;
    this->graphics = graphics;
}

// Getters and setters
string TShirt::getCollar() {
    return this->collar;
}

bool TShirt::getGraphics() {
    return this->graphics;
}

void TShirt::setCollar(string collar) {
    this->collar = collar;
}

void TShirt::setGraphics(bool graphics) {
    this->graphics = graphics;
}

// Display product information
void TShirt::display_product_info() {
    Clothes::display_product_info();
    cout << "Collar: " << this->collar << endl;
    cout << "Graphics: " << this->graphics << endl;
}

// Calculate discount
double TShirt::calculate_discount(string month, int quantity) {
    if (month == "January" && quantity >= 5) {
        return this->getPrice() * 0.1;
    } else {
        return this->getPrice() * 0.05;
    }
}

// Save product information
void TShirt::save(ostream& out) const {
    out << "tshirt" << "\n";
    Clothes::save(out);
    out << collar << "," << graphics << "\n";
}

// Load product information
void TShirt::load(istream& in) {
    Clothes::load(in);

    string line;
    getline(in, line, '\n');
    stringstream ss(line);

    string token;
    getline(ss, token, ',');
    collar = token;

    getline(ss, token, ',');
    graphics = stoi(token);
}

File: Task2A_B/task2a_b/src/main/java/com/cps2004/TShirt.java
package com.cps2004;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public class TShirt extends Clothes {
    private String collar;
    private boolean graphics;

    // Default constructor
    public TShirt() {
        super();
        this.collar = "";
        this.graphics = false;
    }

    // Parameterized constructor
    public TShirt(String name, int id, Supplier supplier, int price, int volume, String size, String brand, String material, String collar, boolean graphics) {
        super(name, id, supplier, price, volume, size, brand, material);
        this.collar = collar;
        this.graphics = graphics;
    }

    // Getters and setters
    public String getCollar() {
        return collar;
    }

    public boolean getGraphics() {
        return graphics;
    }

    public void setCollar(String collar) {
        this.collar = collar;
    }

    public void setGraphics(boolean graphics) {
        this.graphics = graphics;
    }

    @Override
    // Calculate discounted price
    public double calculate_discount(String month, int quantity) {
        if (month.equals("January") && quantity >= 5) {
            return 0.1 * getPrice();
        } else {
            return 0.05 * getPrice();
        }
    }

    // Display product information
    public void display_product_info() {
        super.display_product_info();
        System.out.println("Product collar: " + collar);
        System.out.println("Product graphics: " + graphics);
    }

    // Save product information to file
    public void save(ObjectNode newItem) {
        super.save(newItem);
        newItem.put("collar", this.collar);
        newItem.put("graphics", this.graphics);
    }

    // Load product information from file
    public void load(JsonNode node) {
        super.load(node);
        collar = node.get("collar").asText();
        graphics = node.get("graphics").asBoolean();
    }
}
